### PR TITLE
Fix NPE during removal of VM from Network

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR fixes NPE noticed during removal of  VM from a network (KVM) - leading to test failures noticed in test_vpc_redundant.py
```
2021-06-23 00:17:28,726 WARN  [o.a.c.e.o.NetworkOrchestrator] (API-Job-Executor-87:ctx-86b1bf56 job-3057 ctx-cee7aba3) (logid:355b434c) Unable to complete shutdown of the network elements due to element: VpcVirtualRouter
java.lang.NullPointerException
	at com.cloud.hypervisor.HypervisorGuruBase.toVirtualMachineTO(HypervisorGuruBase.java:183)
	at com.cloud.hypervisor.KVMGuru.implement(KVMGuru.java:112)
	at com.cloud.vm.VirtualMachineManagerImpl.orchestrateRemoveVmFromNetwork(VirtualMachineManagerImpl.java:4171)
	at com.cloud.vm.VirtualMachineManagerImpl.removeVmFromNetwork(VirtualMachineManagerImpl.java:4156)
	at com.cloud.network.router.VpcVirtualNetworkApplianceManagerImpl.removeVpcRouterFromGuestNetwork(VpcVirtualNetworkApplianceManagerImpl.java:207)
	at jdk.internal.reflect.GeneratedMethodAccessor942.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
	at com.sun.proxy.$Proxy250.removeVpcRouterFromGuestNetwork(Unknown Source)
	at com.cloud.network.element.VpcVirtualRouterElement.shutdown(VpcVirtualRouterElement.java:302)
	at org.apache.cloudstack.engine.orchestration.NetworkOrchestrator.shutdownNetworkElementsAndResources(NetworkOrchestrator.java:2949)
	at org.apache.cloudstack.engine.orchestration.NetworkOrchestrator.shutdownNetwork(NetworkOrchestrator.java:2856)
	at org.apache.cloudstack.engine.orchestration.NetworkOrchestrator.destroyNetwork(NetworkOrchestrator.java:3038)
	at com.cloud.network.NetworkServiceImpl.deleteNetwork(NetworkServiceImpl.java:2010)
	at jdk.internal.reflect.GeneratedMethodAccessor682.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.apache.cloudstack.network.contrail.management.EventUtils$EventInterceptor.invoke(EventUtils.java:107)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at com.cloud.event.ActionEventInterceptor.invoke(ActionEventInterceptor.java:51)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
	at com.sun.proxy.$Proxy169.deleteNetwork(Unknown Source)
	at org.apache.cloudstack.api.command.user.network.DeleteNetworkCmd.execute(DeleteNetworkCmd.java:80)
	at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:156)
	at com.cloud.api.ApiAsyncJobDispatcher.runJob(ApiAsyncJobDispatcher.java:108)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:620)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:48)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:45)
	at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:568)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
2021-06-23 00:17:28,737 DEBUG [o.a.c.e.o.NetworkOrchestrator] (API-Job-Executor-87:ctx-86b1bf56 job-3057 ctx-cee7aba3) (logid:355b434c) Lock is released for network Ntwk[336|Guest|73] as a part of network shutdown
2021-06-23 00:17:28,738 DEBUG [o.a.c.e.o.NetworkOrchestrator] (API-Job-Executor-87:ctx-86b1bf56 job-3057 ctx-cee7aba3) (logid:355b434c) Network is not not in the correct state to be destroyed: Shutdown

```

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
code lgtm

ps: it looks not be related to trillian test failures.

**Comment by Pearl1594** [comment]
> code lgtm
> 
> ps: it looks not be related to trillian test failures.

You may be right @weizhouapache, however, I did come across these failures while running the tests and they seemed to be failing during cleanup - however they sometimes also fail to due failure to SSH into instance.

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
> > code lgtm
> > ps: it looks not be related to trillian test failures.
> 
> You may be right @weizhouapache, however, I did come across these failures while running the tests and they seemed to be failing during cleanup - however they sometimes also fail to due failure to SSH into instance.

@Pearl1594 there are same test failures with some prs for 4.15 which does not have this issue.

**Comment by Pearl1594** [comment]
Got it @weizhouapache  - thanks. I'll update the PR description.

**Comment by Pearl1594** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@Pearl1594 do you have a reproduction-/test scheme for us?

**Comment by Pearl1594** [comment]
@DaanHoogland I noticed this exception mainly during the run of the test_vpc_redundant.py smoke test - however, it can be seen occurring at multiple instances on going through the management server logs of the test run.  Another way to validate would be to:
1. Create a VM
2. Attach the VM to another network
3. Shutdown the instance and then attempt removing one of the nics

**Comment by DaanHoogland** [comment]
> @DaanHoogland I noticed this exception mainly during the run of the test_vpc_redundant.py smoke test - however, it can be seen occurring at multiple instances on going through the management server logs of the test run. Another way to validate would be to:
> 
>     1. Create a VM
> 
>     2. Attach the VM to another network
> 
>     3. Shutdown the instance and then attempt removing one of the nics

tnx (I've seen those too lately ;) )

File: server/src/main/java/com/cloud/hypervisor/HypervisorGuruBase.java [hunk 1]
```diff
@@ -179,9 +179,13 @@ protected VirtualMachineTO toVirtualMachineTO(VirtualMachineProfile vmProfile) {
         ServiceOffering offering = _serviceOfferingDao.findById(vmProfile.getId(), vmProfile.getServiceOfferingId());
         VirtualMachine vm = vmProfile.getVirtualMachine();
         HostVO host = hostDao.findById(vm.getHostId());
+        boolean divideMemoryByOverprovisioning = true;
+        boolean divideCpuByOverprovisioning = true;
 
-        boolean divideMemoryByOverprovisioning = VmMinMemoryEqualsMemoryDividedByMemOverprovisioningFactor.valueIn(host.getClusterId());
-        boolean divideCpuByOverprovisioning = VmMinCpuSpeedEqualsCpuSpeedDividedByCpuOverprovisioningFactor.valueIn(host.getClusterId());
+        if (host != null) {
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/hypervisor/HypervisorGuruBase.java:185]: nit - should we check both host != null and host.getClusterId != null? (that may not be necessary but for writing defensive code)?

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/hypervisor/HypervisorGuruBase.java:185]: host.getClusterId() is not referred, only passed. the check should be up to the called method and the method constructing the host object (i think!)

Comment (Pearl1594) [reply server/src/main/java/com/cloud/hypervisor/HypervisorGuruBase.java:185]: @rhtyd  As @DaanHoogland mentioned, the check for validating if the passed argument is null is done by valueIn() (In ConfigKey class) hence there's no need for a check at this point.

**Review by GabrielBrascher — APPROVED** [review]

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan test

**Pull request merged**